Template generics (#2200)
carbon-language/carbon-lang: proposals/p2200.md

    has different return types for certain methods. Anything that can vary in an
    API must be explicitly marked using associated types of an interface, as is
    described in the
    ["parameterized type specialization" design](/docs/design/generics/details.md#specialization).
-   Constraints on a Carbon template type affect how lookup is done into that
    type, as described in
    [proposal #989](https://github.com/carbon-language/carbon-lang/pull/989).

## Details

### Syntax

Template generic bindings are declared using the `template` keyword in addition
to the `:!` of all generic bindings. All generic bindings can only be used in
`let` context to produce r-values, not in a `var` context to produce l-values.
This includes `let` declarations, as in:

```carbon
// `N` is a constant that may be used in types.
let template N:! i64 = 4;
var my_array: [u8; N] = (255, 128, 64, 255);
```

Function parameters also default to a `let` context and may use `template`:

```carbon
// `U` is a templated type parameter that must be specified
// explicitly by the caller.
fn Cast[template T:! Type](x: T, template U:! Type) -> U {

>   return x as U;

Reviewer: Maybe this would help a reader understand how this contrasts templates to checked generics:
```suggestion
  // OK, check for `T is As(U)` delayed until values of `T` and `U` are known.
  return x as U;
```

The `auto` keyword may also be used to omit the return type, as specified in
[#826: Function return type inference](https://github.com/carbon-language/carbon-lang/pull/826).
This feature is being reconsidered in the open question-for-leads issue
[#1008: How much complexity should we invest in deduced/inferred return types?](https://github.com/carbon-language/carbon-lang/issues/1008).

The semantics of `let x:! auto = ...` is the subject of open question-for-leads
issue
[#996: Generic `let` with `auto`?](https://github.com/carbon-language/carbon-lang/issues/996).

### Template constraints

Template constraints have already been introduced in proposal
[#818: Constraints for generics](https://github.com/carbon-language/carbon-lang/pull/818).
In brief, a `template constraint` declaration is like a `constraint`
declaration, except that it may also contain function and field declarations,
called _structural constraints_. Only types with matching declarations will
satisfying the template constraint. Note that the declarations matching the
structural constraints must be in the type declaration itself, not in an

> external impl.

Reviewer: I assume "in the type declaration itself" includes defaulted methods from internally-implemented interfaces, as well as members from base classes. I'm not sure if we have good terminology here, but something like this seems like it could work:

```suggestion
structural constraints must be found by member lookups in the type. It is not
sufficient for them to be declared only in an external impl.
```

(I think "not in an external impl" could be read as saying that declarations are fully disqualified if they come from external impls, which I don't think is intended -- if the name is aliased into the type, that should be enough, I think.)

```carbon
interface A { fn F[me: Self](); }
interface B { fn F[me: Self](); }
class C { }
external impl C as A;
external impl C as B;
template constraint HasF {
  fn F[me: Self]();
}
fn G[template T:! HasF](x: T);
var y: C = {};
// Can't call `G` with with `y` since it doesn't have any internal
// implementation of a method `F` satisfying `HasF`, even though `C`
// externally implements both `A` and `B` with such an `F`. May
// define an adapter for `C` to get a type that implements `HasF`,
// with `A.F`, `B.F`, or some other definition.
```

This was discussed in
[#generics-and-templates on 2022-09-20](https://discord.com/channels/655572317891461132/941071822756143115/1021903925613449316).

Structural constraints do not affect [name lookup](#name-lookup) into template
type parameters. They guarantee that a name will be available in the type, but
don't change the outcome.

```carbon
template constraint HasF {
  fn F[me: Self]();
}
class C {
  fn F[me: Self]();
}
fn G[template T:! HasF](x: T) {
  x.F();
}
var y: C = {};
// Call to `F` inside `G` is not ambiguous since
// `C.F` and `HasF.F` refer to the same function.
G(y);

> ```

Reviewer: Example for the comment I made above about "not in an external impl".
````suggestion
class D extends C {
  alias F = C.(A.F);
}
// OK, `z.(HasF.F)` will resolve to `z.(C.(A.F))`.
fn Run(z: D) { G(z); }
```
````

Whether template constraints may be used as constraints on checked-generic
parameters is being considered in question-for-leads issue
[#2153: Generics calling templates](https://github.com/carbon-language/carbon-lang/issues/2153).
Even if we allow a checked-generic parameter to use a template constraint, we
want to focus checked generics on semantic properties encapsulated in
interfaces, not structural properties tested by template constraints. So we

> would not allow lookup into checked-generic type to find type members outside of

Reviewer: ```suggestion
would not allow lookup into a checked-generic type to find type members outside of
```

type parameters. They guarantee that a name will be available in the type, but
don't change the outcome.

```carbon
template constraint HasF {
  fn F[me: Self]();
}
class C {
  fn F[me: Self]();
}
fn G[template T:! HasF](x: T) {
  x.F();
}
var y: C = {};
// Call to `F` inside `G` is not ambiguous since
// `C.F` and `HasF.F` refer to the same function.
G(y);
```

Whether template constraints may be used as constraints on checked-generic
parameters is being considered in question-for-leads issue
[#2153: Generics calling templates](https://github.com/carbon-language/carbon-lang/issues/2153).
Even if we allow a checked-generic parameter to use a template constraint, we
want to focus checked generics on semantic properties encapsulated in
interfaces, not structural properties tested by template constraints. So we
would not allow lookup into checked-generic type to find type members outside of
an interface:

```carbon
template constraint HasF {
  fn F[me: Self]();
}
// ❓ If we allow a checked generic to use a template
// constraint, as in:
fn H[T:! HasF](x: T) {
  // We still will not support calling `F` on `x`:
  // ❌ x.F();
}
```

> These members would only be found using a template type parameter.

Reviewer: Presumably one can write:
```
template constraint TemplateHasF {
  fn F[me: Self]();
}
interface MyHasF {
  fn F[me: Self]();
}
final impl forall [template T:! TemplateHasF] T as MyHasF {
  alias F = T.F;
}
```
... and get the effect of a template constraint that can be used in a checked generic without saying `template`:
```
fn MyH[T:! MyHasF](x: T) {
  // OK, member of interface `MyHasF`.
  x.F();
}
```
I'm not sure what I conclude from this. Your template constraint `HasF` seems pretty similar to my interface `MyHasF`, but they're clearly not the same: mine supports providing an `impl` for a type that has no `F` member name.

I guess one thing that's not clear to me is: are the members in a template constraint *only* a predicate on the type, used to determine whether the constraint is satisfied? Or are they also used in type-checking uses of types with the constraint, in the case where we constrain a template type parameter:
```
fn TemplateH[template T:! HasF](x: T) {
  // OK, with instantiation errors? Or do we reject this when type-checking the template?
  x.F(4);
}
```
If that's invalid, because we do use `HasF` when type-checking a template, it seems a little odd to me that we wouldn't allow the call when type-checking a checked generic like `H`.

Are there things that you can do in a template constraint that you can't do in an interface, so that the translation from `HasF` to `MyHasF` above can't be done, and so that we can't type-check a function definition like `H`?

Author: 1) We definitely want to support template implementation of interfaces, and we do want them to be similar semantically to template constraints so you can safely transition off of template constraints. This is part of the "Transition from C++ templates to Carbon checked generics" story, see the "To interface constraints" section.

2) I'm not sure I can answer the questions about early checking of the `x.F(4)` call with templates until we work out the overloading story, which interacts with this question. Maybe `T` has another overload of `F` such that the call would be allowed?

3) The point here is that it would be sound to support structural constraints with checked generics, much like Go does, but that we are intentionally choosing not to. This is absolutely something you might disagree with or that we decide to change in the future, it is not forced based on our other decisions. This is under the "features that pierce abstraction boundaries that we want to discourage for software engineering reasons or at least mark when they are in use" umbrella. This is here as part of defining the differences between checked and template generics.

    function.

In either case, the compiler will give an error if the interface is not
implemented for some types before the step is finished.

An example of the second approach, starting with a templated function with a
structural constraint:

```carbon
template constraint HasF {
  fn F[me: Self]();
}

fn G[template T:! HasF](x: T) {
  x.F();
}

class C {
  fn F[me: Self]();
}
var y: C = {};
G(y);
```

First, a new interface is created with a blanket implementation and the
function's constraints are updated to use it instead. Calls in the function body
should be qualified to avoid ambiguity errors.

```carbon
template constraint HasF {
  fn F[me: Self]();
}

// New interface
interface NewF {
  fn DoF[me: Self]();
}

// Blanket implementation
external impl forall [template T:! HasF] T as NewF {
  // If the functions are identical, can instead do:
  //    alias DoF = T.F;
  fn DoF[me: Self]() {
    me.F();
    // Or: me.(T.F)();
  }
}

// Changed constraint
fn G[template T:! NewF](x: T) {
  // Call function from interface
  x.(NewF.DoF)();
  // Could use `x.DoF();` instead, but that will
  // give a compile error if `T` has a definition

>   // for `DoF` in addition to the one in `NewF`.

Reviewer: If we use the name `F` rather than `DoF` in `NewF`, and use `alias F = T.F;` in the impl, can we avoid this ambiguity problem and use `x.DoF` here instead? I think both lookups should then always give the same result.

Author: Without qualification, I think there will be opportunities for ambiguity. They might not happen for these types, but other types with a method `F` and/or a specialization of the `NewF` implementation. So I think it would be legal to reuse `F` (though that might make the example less clear), I still think the recommendation here is to include a qualification.

depends on the value of the parameter, not just its type. As an example, whether
two array types are compatible depends on whether they have the same size. With
a symbolic constant sizes, they will only be considered equal if the compiler
can show that the two sizes are always equal symbolically. If the size is a
template parameter, the checking will be delayed until the value of the template
parameter is known.

### Template dependent

Expressions fall under three categories:

-   Expressions that are valid and have meaning determined without knowing the
    value of any template parameter are _not template dependent_.
-   Expressions whose meaning and validity requires knowing the value of a
    template parameter are _template dependent_.
-   Expressions that have a meaning without knowing the value of any template
    parameter, assuming it is valid, but whose validity requires knowing the
    value of a template parameter are _template validity dependent_. These
    expressions can trigger a monomorphization error, but are not considered
    template dependent for purposes of a containing expression.

#### Unqualified member name lookup

> There are two cases when performing unqualified member-name lookup into a

Reviewer: ```suggestion
There are three cases when performing unqualified member-name lookup into a
```

    dependent. This means that template instantiation may fail if `T` has a
    member `G` different than `T.(I.G)`. Assuming it succceeds, though, it will
    definitely have meaning determined by `I.G`.
-   If the member name is not found in the constraint, lookup may still succeed
    once the type is known, so the result is template dependent.
-   If the lookup is ambiguous prior to knowing the value of the type, for
    example if `G` has two distinct meanings in `I`, then the code is invalid.

The value of the expression will be dependent. For example, if `U` is an
associated type of `I`, then `T.U` as an expression is template validity
dependent, but the value of that expression is dependent. The value is not
always needed to perform checking, for example `x.G()` can be checked without
ever determining the value of `x.G` as long as its signature can be determined,
if it is valid.

#### Qualified member name lookup

Adding qualifier to the member name, as in `T.(U.X)`, can make the lookup less
dependent on the template parameter:

-   If `T` is dependent, and `U` is an interface, the value of the expression is

>     dependent, but the type of the expression is not.

Reviewer: Is this always true? We might not know whether `T` is a type or non-type value:

```
interface U {
  fn X[me: Self]();
}
fn F[T:! Type](x: T) {
  // We don't know if `T.F()` is a type or non-type, so we don't know if
  // this is performing instance binding or returning an unbound member name.
  T.F().(U.X)();
}
```

I suppose we could say the type of `T.F().(U.X)` is known either way, and that the fact that `SomeType.(U.X)` is an unbound member name is tracked in some other way. But `SomeType.(U.X)` and `SomeValue.(U.X)` still seem like maybe they should have different types in that the former has a `[me: Self]` parameter and the latter presumably does not -- though this will depend on whether we have function types and how they work, which is not decided yet.

Author: I think you are right, but I'm not entirely sure how to fix it. Should I be separating the lookup from the type and the value?

Author: I've reverted my change here. Under our new ideas for compound lookup, it would assume instance binding.

Reviewer: It'd be clearer if `T` were renamed to something like `x` here, given that we're now assuming we always perform instance binding, so `T` should be an instance, typically not a type.

I don't think the conclusion is quite correct. I think the type of the expression would still be dependent if the type of `U.X` involves `Self`:

```suggestion
-   If `T` is dependent, and `U` is an interface, the value of the expression is
    dependent, but the type of the expression is not dependent unless the type
    of `U.X` involves `Self`.
```

    -   If `T` is known to implement `U`, then the lookup is not dependent. For
        example, there could be a requirement on `T` or a sufficiently general

>         implementation of `U` that includes all possible values of `T`.

Reviewer: If we don't find a final impl, shouldn't the value still be dependent as in the previous bullet? Or does this only mean that the lookup is not template validity dependent in this case?

Author: I've tried to update the text.

    -   Otherwise, the lookup is template validity dependent.

```carbon
interface Hashable {
  let HashType:! Type;
  fn Hash[me: Self]() -> HashType;
}

external impl forall [T:! Serializable] T as Hashable;

fn F[template T:! Serializable](x: T) {
  // `T` is required to implement `Serializable` so this
  // is not dependent.

>   x.(Serializable.Serialize)();

Reviewer: I would expect the value of the lookup to be dependent, but not the type, and for it to not be template validity dependent.

Author: I said it isn't dependent since the value is not used here.

If the validity of an expression requires that an impl exist for a type, and
that can't be determined until the value of a template parameter is known, then
the expression is template validity dependent. For example, in the example from
the previous section, `x.(Printable.Print)()` is valid if `T` implements
`Printable`. This can also occur without a qualified lookup, for example:

```carbon
fn F[T:! Printable](x: T);
fn G[template T:! Type](x: T) {
  // Valid if `T` implements `Printable`, so this
  // expression is template validity dependent.
  F(x);
}
```

The values of members of an impl for a template-dependent type are template
dependent, unless they can be resolved to a not template-dependent expression
using checked-generic impl resolution.

```
final external impl [T:! Type] T* as D
    where .Result = T and .Index = i32;
fn F[T:! Type](p: T*) {
  // `(T*).(D.Index)` uses the final impl of `D` for `T*`,
  // and so equals `i32`, which is not dependent.
  // `(T*).(D.Result)` is recognized as equal to `T`, which
  // is template dependent.
}
```

#### Use of dependent value is dependent

To match the expectations of C++ templates, uses of dependent values are also

> template dependent. For example:

Reviewer: The example below is a bit unusual: we propagate the dependence of the case condition down (and sideways) into everything in the body. The normal case would be propagating dependence upwards, from subexpressions to enclosing expressions. Maybe we should give an example of that case first, and then discuss the `match` case as a special case?

Author: What is an example of the normal case you are thinking of?

Reviewer: For example, if `expr` is a dependent expression, is each of these dependent:
- `(expr)`
- `F(expr)`
- `expr.a` (discussed above)
- `expr + 1`
- `if expr then a else b`
- `if a then expr else b`
- `expr as T`
- `a as expr`

Author: Updated.

the interfaces are changed.

### Validity can depend on value

A templated parameter may be used in ways where the validity of the result
depends on the value of the parameter, not just its type. As an example, whether
two array types are compatible depends on whether they have the same size. With
a symbolic constant sizes, they will only be considered equal if the compiler
can show that the two sizes are always equal symbolically. If the size is a
template parameter, the checking will be delayed until the value of the template
parameter is known.

### Template dependent

Expressions fall under three categories:

-   Expressions that are valid and have meaning determined without knowing the
    value of any template parameter are _not template dependent_.
-   Expressions whose meaning and validity requires knowing the value of a
    template parameter are _template dependent_.
-   Expressions that have a meaning without knowing the value of any template
    parameter, assuming it is valid, but whose validity requires knowing the
    value of a template parameter are _template validity dependent_. These
    expressions can trigger a monomorphization error, but are not considered

>     template dependent for purposes of a containing expression.

Reviewer: It's not clear whether these are properties of the expression and its subexpressions, or of the expression in isolation in some sense. For example, given `(dependent expr).(A.B)`, the full meaning of that expression depends on the meaning of `dependent expr` -- we can't evaluate that expression without evaluating the dependent subexpression -- which makes it seem like that expression should be dependent. But I think the idea is that that expression is _not_ dependent, because we only care about the properties of the expression independent of its subexpressions for the purpose of the definition of _template dependent_.

It's not clear to me whether the same is true for _template validity dependent_, or whether `(dependent expr).(A.B)` would be template validity dependent if `A.B` is, even if we know that the member access expression is always valid regardless of the properties of the dependent subexpression.

Author: How can I phrase this to be clearer? I think I was trying to get at what you were saying with "but are not considered template dependent for purposes of a containing expression". Maybe there is an example that would clarify?

Author: I added a paragraph https://github.com/carbon-language/carbon-lang/pull/2200/commits/1a670dd5e5bf64688956a681023910479cc23c39 , does that address your concerns?

Reviewer: I think it's now clear that "template dependent" refers only to properties of the "immediate" expression and not to its subexpressions. But I still don't know whether template validity dependence takes subexpression validity into account or not. Given:
```
fn G(n: i32) -> i32;
fn F[template T:! Type](x: T) -> i32 { return G(x.y); }
```
... is the expression `G(x.y)` template validity dependent, because its validity depends on whether `T` has a member `y`, or is it not template dependent because we know that it is an rvalue of type `i32` that calls `G` regardless of the properties of `T`?

I don't think this needs to be resolved immediately -- we're not using this notion for much, yet -- but we'll need an answer once we start using this terminology to drive other language rules.

    definitely have meaning determined by `I.G`. There may still be ambiguity
    making the result dependent if it is not known whether `x` is a type and
    `I.G` is a method and so has an implicit `me` parameter.
-   If the member name is not found in the constraint, lookup may still succeed
    once the type is known, so the result is template dependent.
-   If the lookup is ambiguous prior to knowing the value of the type, for
    example if `G` has two distinct meanings in `I`, then the code is invalid.

The value of the expression will be dependent. For example, if `U` is an
associated type of `I`, then `T.U` as an expression is template validity
dependent, but the value of that expression is dependent. The value is not
always needed to perform checking, for example `x.G()` can be checked without
ever determining the value of `x.G` as long as its signature can be determined,
if it is valid.

#### Qualified member name lookup

Adding qualifier to the member name, as in `T.(U.X)`, can make the lookup less
dependent on the template parameter:

-   If `T` is dependent, and `U` is an interface, the value of the expression is
    dependent, but the type of the expression is not.

>     -   If `T` is known to implement `U`, then the lookup is not dependent, see

Reviewer: I think this should now be... "If the type of `T` is known to implement `U`, ..." after #2360.

Author: Updated. I've also renamed `T` to `x` (and `X` to `V`).

in addition to
[#1371: Is `let` referentially transparent?](https://github.com/carbon-language/carbon-lang/issues/1371).

**Note:** Exactly which expressions in terms of constants result in constants is
an open question that is not resolved by this proposal. In particular, which
function calls will be evaluated at compile time is not yet specified. See
[future work](#which-expressions-will-be-evaluated-at-compile-time).

### `auto`

The `auto` keyword is a shortcut for an unnamed templated type, as in:

```carbon
// Type of `x` is the same as the return type of function `F`.
let x: auto = F();
```

This was first added to Carbon in proposal
[#553: Generic details part 1](https://github.com/carbon-language/carbon-lang/pull/553)
and further specified by open proposal
[#2188: Pattern matching syntax and semantics](https://github.com/carbon-language/carbon-lang/pull/2188).

The `auto` keyword may also be used to omit the return type, as specified in
[#826: Function return type inference](https://github.com/carbon-language/carbon-lang/pull/826).
This feature is being reconsidered in the open question-for-leads issue

> [#1008: How much complexity should we invest in deduced/inferred return types?](https://github.com/carbon-language/carbon-lang/issues/1008).

Reviewer: This issue has been resolved now.

The semantics of `let x:! auto = ...` is the subject of open question-for-leads
issue
[#996: Generic `let` with `auto`?](https://github.com/carbon-language/carbon-lang/issues/996).

### Template constraints

Template constraints have already been introduced in proposal
[#818: Constraints for generics](https://github.com/carbon-language/carbon-lang/pull/818).
In brief, a `template constraint` declaration is like a `constraint`
declaration, except that it may also contain function and field declarations,
called _structural constraints_. Only types with matching declarations will

> satisfying the template constraint. Note that the declarations matching the

Reviewer: ```suggestion
satisfy the template constraint. Note that the declarations matching the
```

interface NewF {
  fn DoF[me: Self]();
}

// Blanket implementation
external impl forall [template T:! HasF] T as NewF {
  // If the functions are identical, can instead do:
  //    alias DoF = T.F;
  fn DoF[me: Self]() {
    me.F();
    // Or: me.(T.F)();
  }
}

// Changed constraint
fn G[template T:! NewF](x: T) {
  // Call function from interface
  x.(NewF.DoF)();
  // Could use `x.DoF();` instead, but that will
  // give a compile error if `T` has a definition
  // for `DoF` in addition to the one in `NewF`.
}

class C {
  fn F[me: Self]();
}
var y: C = {};
// Still works since `C` implements `NewF`
// from blanket implementation.
G(y);
```

Then the interface is implemented for types used as parameters:

```carbon
// ...
class C {
  impl as NewF {
    // `NewF.DoF` will be called by `G`, not `C.F`.
    fn DoF[me: Self]();
  }
  // No longer needed: fn F[me: Self]();
}
// ...
```

Once all types have implemented the new interface, the blanket implementation
can be removed:

```carbon
// Template constraint `HasF` no longer needed.

// New interface
interface NewF {
  fn DoF[me: Self]();
}

// Blanket implementation no longer needed.

fn G[template T:! NewF](x: T) {

>   x.DoF();

Reviewer: ```suggestion
  x.(NewF.DoF)();
```
We're not removing the `NewF.` qualification until the next step, I think.

dependent, but the value of that expression is dependent. The value is not
always needed to perform checking, for example `x.G()` can be checked without
ever determining the value of `x.G` as long as its signature can be determined,
if it is valid.

#### Compound member access

Adding qualifier to the member name, as in `x.(U.V)`, can make the lookup less
dependent on the template parameter:

-   If `x` is dependent, and `U` is an interface, the value of the expression is
    dependent, but the type of the expression is not dependent unless the type
    of `U.V` involves `Self`. The lookup itself follows proposal
    [#2360](https://github.com/carbon-language/carbon-lang/pull/2360):

    -   If `U.V` is an _instance_ member, and the type of `x` is known to
        implement `U`, then the lookup is not dependent. For example, there
        could be a requirement on `x` or a sufficiently general implementation
        of `U` that includes all possible types of `x`. The resulting value is
        dependent unless the implementation of `U` is `final`, see
        [the impl lookup section](#impl-lookup).
    -   Otherwise, the lookup is template validity dependent.

    ```carbon

>     interface Hashable {

Reviewer: ```suggestion
    interface Serializable {
      fn Serialize[me: Self]();
    }
    
    interface Printable {
      fn Print[me: Self]();
    }

    interface Hashable {
```
Given that the example below depends on whether `Serialize` is an instance member, and also on the return type of `Serialize` (specifically, if it involves `Self`), it seems useful to define `Serializable` here. Similarly for `Printable`.

`Printable`. This can also occur without a qualified lookup, for example:

```carbon
fn F[T:! Printable](x: T);
fn G[template T:! Type](x: T) {
  // Valid if `T` implements `Printable`, so this
  // expression is template validity dependent.
  F(x);
}
```

The values of members of an impl for a template-dependent type are template
dependent, unless they can be resolved to a not template-dependent expression
using checked-generic impl resolution.

```
final external impl [T:! Type] T* as D
    where .Result = T and .Index = i32;
fn F[T:! Type](p: T*) {
  // `(T*).(D.Index)` uses the final impl of `D` for `T*`,
  // and so equals `i32`, which is not dependent.
  // `(T*).(D.Result)` is recognized as equal to `T`, which
  // is template dependent.
}
```

#### Use of dependent value is dependent

To match the expectations of C++ templates, uses of dependent values are also
template dependent, propagating dependence from subexpressions to enclosing
expressions. For example, if `expr` is a dependent expression, each of these is
dependent:

-   `(expr)`
-   `F(expr)`
-   `expr + 1`
-   `if expr then a else b`
-   `if a then expr else b`
-   `expr as T`

> -   `a as expr`

Reviewer: Depending on how we make use of the term _dependent_ in other language rules, I think we may want to revisit some of these decisions (eg, `if expr then a else b` and `expr as T` seem like cases where we can say that `expr` being dependent doesn't make the enclosing expression dependent), but I'm happy with using this conservative rule for now.

Author: I added short paragraph with this. Or I'd be happy to mark those cases as not dependent.

    -   [Template dependent](#template-dependent)
        -   [Simple member access](#simple-member-access)
        -   [Compound member access](#compound-member-access)
        -   [Impl lookup](#impl-lookup)
        -   [Use of dependent value is dependent](#use-of-dependent-value-is-dependent)
-   [Rationale](#rationale)
-   [Alternatives considered](#alternatives-considered)
    -   [Only checked generics](#only-checked-generics)
    -   [SFINAE](#sfinae)
    -   [Ad hoc API specialization](#ad-hoc-api-specialization)
    -   [Value phase of bindings determined by initializer](#value-phase-of-bindings-determined-by-initializer)
    -   [Simpler template dependent rules that delayed more checking](#simpler-template-dependent-rules-that-delayed-more-checking)
-   [Future work](#future-work)
    -   [Expanded template constraints](#expanded-template-constraints)
    -   [Predicates: constraints on values](#predicates-constraints-on-values)
    -   [Checked generics calling templates](#checked-generics-calling-templates)
    -   [Which expressions will be evaluated at compile time](#which-expressions-will-be-evaluated-at-compile-time)

<!-- tocstop -->

## Abstract

> Add template generics, with optional constraints but no SFINAE, to Carbon.

Reviewer 2: ```suggestion
Add template generics, with optional constraints but no [SFINAE](https://en.cppreference.com/w/cpp/language/sfinae), to Carbon.
```

(Or a forward link into the terminology section which I see has a great definition here, mostly just trying to solve the forward-reference.)

        -   [To template Carbon with structural constraints](#to-template-carbon-with-structural-constraints)
        -   [To interface constraints](#to-interface-constraints)
        -   [To checked generic](#to-checked-generic)
    -   [Validity can depend on value](#validity-can-depend-on-value)
-   [Rationale](#rationale)
-   [Alternatives considered](#alternatives-considered)
    -   [Only checked generics](#only-checked-generics)
    -   [SFINAE](#sfinae)
    -   [Ad hoc API specialization](#ad-hoc-api-specialization)
    -   [Value phase of bindings determined by initializer](#value-phase-of-bindings-determined-by-initializer)
-   [Future work](#future-work)
    -   [Expanded template constraints](#expanded-template-constraints)
    -   [Predicates: constraints on values](#predicates-constraints-on-values)
    -   [Checked generics calling templates](#checked-generics-calling-templates)
    -   [Which expressions will be evaluated at compile time](#which-expressions-will-be-evaluated-at-compile-time)

<!-- tocstop -->

## Abstract

Add template generics, with optional constraints but no SFINAE, to Carbon.

> Template generics allows the compiler postpone type checking of expressions

Reviewer: ```suggestion
Template generics allows the compiler to postpone type checking of expressions
```

> Template generics allows the compiler postpone type checking of expressions

Reviewer: Not sure I have a better word to suggest, but "allows" seems misleading here. This is more "requires" I think?

But more broadly, I wonder if we should put the emphasis differently, on the user instead of the compiler, maybe something like: template generics allows a developer to write code against the interface of a family of types that share a common API, even if the specific types are not known where the template is defined and the API is not formally defined.

    and template generics, and definitions for terms like _instantiation_.
-   Proposal
    [#553: Generic details part 1](https://github.com/carbon-language/carbon-lang/pull/553)
    defined `auto` as a template construct, and described how templates do not
    require constraints to find member names.
-   Question-for-leads issue
    [#565: Generic syntax to replace provisional `$`s](https://github.com/carbon-language/carbon-lang/issues/565)
    implemented in proposal
    [#676: `:!` generic syntax](https://github.com/carbon-language/carbon-lang/pull/676)
    defined the syntax for template bindings.
-   Proposal
    [#731: Generics details 2: adapters, associated types, parameterized interfaces](https://github.com/carbon-language/carbon-lang/pull/731)
    included that template values may be passed to generic parameters.
-   Proposal
    [#818: Constraints for generics](https://github.com/carbon-language/carbon-lang/pull/818)
    included `template constraint` to defined named constraints with fewer
    restrictions for use with template parameters.
-   Proposal
    [#875: Principle: information accumulation](https://github.com/carbon-language/carbon-lang/pull/875)

>     considered how the principle benefited and was impacted by templates.

Reviewer: Just so you can share in my surprise: "benefited" is spelled differently in British English ("benefitted").

    advertised that Carbon would support templates for "seamless C++ interop."
-   Terminology was updated in proposal
    [#2138: Checked and template generic terminology](https://github.com/carbon-language/carbon-lang/pull/2138).

TODO: Update if proposal
[#2188: Pattern matching syntax and semantics](https://github.com/carbon-language/carbon-lang/pull/2188)
is accepted first.

## Terminology

-   A _dependent_ name or expression is one whose meaning depends on, that is
    varies with, some (checked or template) generic parameter. This is
    consistent with the meaning of this term in
    [C++](https://en.cppreference.com/w/cpp/language/dependent_name), but is
    different from
    ["dependent types"](https://en.wikipedia.org/wiki/Dependent_type).
-   [_Instantiation_](/docs/design/generics/terminology.md#instantiation),
    _substitution_, or
    [_monomorphizaton_](https://en.wikipedia.org/wiki/Monomorphization) is the
    process of duplicating the implementation of a function and then
    substituting in the values of any (checked or template) generic arguments.
-   Errors that are only detected once the argument value from the call site are

>     known as _monomorphization errors_. These mostly occur in expressions

Reviewer: Missing some words in the first sentence. "... call site is known are known as ..." would be correct but reads very strangely! Maybe:

```suggestion
-   Errors that are only detected once the argument value from the call site is
    known are called _monomorphization errors_. These mostly occur in expressions
```

    dependent on some template parameter, but can also occur for other reasons
    like hitting an implementation limit.
-   _SFINAE_ stands for "Substitution failure is not an error", which is the
    policy in C++, see
    [cppreference](https://en.cppreference.com/w/cpp/language/sfinae),
    [wikipedia](https://en.wikipedia.org/wiki/Substitution_failure_is_not_an_error).
    It means that functions from an overload set with monorphization errors, or
    "substitution failure," are simply ignored instead of causing compilation to

>     fail.

Reviewer: ```suggestion
    It means that functions from an overload set with monorphization errors, or
    "substitution failure," within their signatures are simply ignored instead of causing compilation to
    fail.
```

Probably worth calling out here that C++ only applies SFINAE within the "immediate context" of the function declaration, which isn't actually formally defined (!) but excludes function bodies and generally any other declaration / definition whose instantiation is triggered (but includes alias templates because, er, reasons?).

## Proposal

We propose that template generics are included as an official feature of Carbon.

In many ways, template generic parameters work like checked generic parameters.
The following are true for any kind of generic parameter:

-   The value passed to a generic parameter must be able to be evaluated at
    compile time.
-   Generic parameters may have constraints that will be enforced by the
    compiler on the value supplied by the caller.
-   The compiler may choose to generate multiple copies of a generic function
    for different values of the generic parameters.

The main differences between checked and templated generics are:

-   Member lookup into a templated type looks in both the actual type value
    provided by the caller in addition to in any constraints on that type see

>     [proposal #989](https://github.com/carbon-language/carbon-lang/pull/989).

Reviewer: ```suggestion
-   Member lookup into a templated type looks in the actual type value
    provided by the caller in addition to in any constraints on that type; see
    [proposal #989](https://github.com/carbon-language/carbon-lang/pull/989).
```

-   A templated parameter may be used in ways where the validity of the result
    depends on the value of the parameter, not just its type.

As a consequence of these differences, type checking of any expression dependent
on a templated parameter may not be completed until its value is known. In
addition, templated generics support branching on the value of a templated type.

In contrast with C++ templates, with Carbon template generics:

-   Substitution failure is an error. In C++, [the SFINAE rule](#terminology)
    will skip functions in overload resolution that fail to instantiate.
    Instead, Carbon template parameters use constraints to control when the
    function is available.
-   Carbon template specialization does not allow ad hoc changes to the API of
    the function or type being specialized, only its implementation. This is in
    contrast to C++, where
    [C++'s `std::vector<bool>`](https://en.cppreference.com/w/cpp/container/vector_bool)

>     has different return types for certain methods. Anything that can vary in an

Reviewer: Also, `std::vector<bool>` has [additional methods](https://en.cppreference.com/w/cpp/container/vector_bool/flip)!

Author: While true, I don't think that is as much of a problem since code written against the `std::vector<T>` API won't call those additional methods and will continue to work. A different return type for methods being called is much more serious in my mind.

    API must be explicitly marked using associated types of an interface, as is
    described in the
    ["parameterized type specialization" design](/docs/design/generics/details.md#specialization).
-   Constraints on a Carbon template type affect how lookup is done into that
    type, as described in
    [proposal #989](https://github.com/carbon-language/carbon-lang/pull/989).

## Details

### Syntax

Template generic bindings are declared using the `template` keyword in addition
to the `:!` of all generic bindings. All generic bindings can only be used in
`let` context to produce r-values, not in a `var` context to produce l-values.

> This includes `let` declarations, as in:

Reviewer: It took me a couple of reads to understand what the "All generic bindings can only ..." sentence was saying. A negative example would help, eg:
```
// Error, can't declare a variable constant.
var N:! i64 = 4;
```

Author: Rearranged to include an example.
